### PR TITLE
[CALCITE-2892] Add the JSON_KEYS function

### PR DESCRIPTION
**Pull request by XuQianJin-Stars**
[JSON_KEYS({{json_doc[, _{{path}}_])}}|https://dev.mysql.com/doc/refman/5.7/en/json-search-functions.html#function_json-keys]

Returns the keys from the top-level value of a JSON object as a JSON array, or, if a path argument is given, the top-level keys from the selected path. Returns NULL if any argument is NULL, the json_doc argument is not an object, or path, if given, does not locate an object. An error occurs if the json_doc_argument is not a valid JSON document or the path_ argument is not a valid path expression or contains a {} or * wildcard.

The result array is empty if the selected object is empty. If the top-level value has nested subobjects, the return value does not include keys from those subobjects.

Example SQL:
```sql
SELECT JSON_KEYS(v) AS c1
,JSON_KEYS(v, 'lax $.a') AS c2
,JSON_KEYS(v, '$.b') AS c3
,JSON_KEYS(v, 'strict $.a[0]') AS c4
,JSON_KEYS(v, 'strict $.a[1]') AS c5
FROM (VALUES ('{"a": [10, true],"b": {"c": 30}}')) AS t(v)
LIMIT 10;
```
Result:

c1 | c2 | c3 | c4 | c5
-- | -- | -- | -- | --
["a", "b"] | [NULL] | ["c"] | [NULL] | [NULL]

### CODE REVIEW

**Comment by michaelmior** [comment]
@XuQianJin-Stars Thanks for this! Could you check out the test failures?

**Comment by michaelmior** [comment]
@XuQianJin-Stars Could you explain what was invalid about that test that meant it should just be deleted?

**Comment by XuQianJin-Stars** [comment]
hi @michaelmior Thank you very much. I will add and improve this test as soon as possible. I am busy recently, this problem and the PR of json_length. I'll handle it together.

**Comment by XuQianJin-Stars** [comment]
hi @michaelmior @zhztheplayer I have updated this PR. Can you continue to review this PR?
best
qianjin

**Comment by zhztheplayer** [comment]
> hi @michaelmior @zhztheplayer I have updated this PR. Can you continue to review this PR?
> best
> qianjin

Thanks @XuQianJin-Stars for the contribution! We have made changes on JSON functions' documentation in https://github.com/apache/calcite/commit/9bce62c6b6abd13d551d0eed1f1e14b024bbb2ae. Could you please solve the conflicts accordingly first?

**Comment by XuQianJin-Stars** [comment]
hi @michaelmior @zhztheplayer I have updated this PR. Can you continue to review this PR?
best
qianjin

File: babel/src/main/codegen/config.fmpp [hunk 1]
```diff
@@ -557,6 +557,10 @@ data: {
         "JSON_ARRAYAGG",
         "JSON_EXISTS",
         "JSON_VALUE",
+        "JSON_TYPE",
```
Comment (zhztheplayer) [inline babel/src/main/codegen/config.fmpp]: Why 4 keywords added here?

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 2]
```diff
@@ -876,6 +876,9 @@
 
   @BaseMessage("Not a valid input for JSON_LENGTH: ''{0}''")
   ExInst<CalciteException> invalidInputForJsonLength(String value);
+
+  @BaseMessage("Unknown JSON keys in JSON_KEYS function, and the object is: ''{0}''")
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java]: Not a valid input for JSON_KEYS: ''{0}''

File: babel/src/main/codegen/config.fmpp (hunk 1 above)
Comment (XuQianJin-Stars) [reply babel/src/main/codegen/config.fmpp]: hi @zhztheplayer Because I have missed something that was not added before.

Comment (zhztheplayer) [reply babel/src/main/codegen/config.fmpp]: I see, thanks for the explanation.

Comment (zhztheplayer) [reply babel/src/main/codegen/config.fmpp]: And I see the format of non-reserved keyword list in babel parser is a little messy. I'll clean it a bit on top of your work.

Comment (XuQianJin-Stars) [reply babel/src/main/codegen/config.fmpp]: Well, thank you very much.

**Pull request closed without merging**